### PR TITLE
feat: onboard comparator logic

### PR DESCRIPTION
**Pull request by utsabc**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/services/comparator.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+import ServiceSelector from '../helpers/serviceSelector';
+import tags from '../v0/util/tags';
+import stats from '../util/stats';
+import logger from '../logger';
+import { CommonUtils } from '../util/common';
+import { PlatformError } from '../v0/util/errorTypes';
+const NS_PER_SEC = 1e9;
+
+export default class ComparatorService {
+  private static getTestThreshold(destination: Destination) {
```
Comment (koladilip) [inline src/services/comparator.ts:12]: getCdkTestThreshold

File: src/services/comparator.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    return destination.DestinationDefinition?.Config['cdkV2TestThreshold'] || 0;
+  }
+
+
+  public static async compareDestinationServiceAtProc(
```
Comment (koladilip) [inline src/services/comparator.ts:17]: compareDestinationServiceAtProc name appears that it is processor but inside there is logic for everything

File: src/services/comparator.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    const { events } = input;
+    const { version, destination }: { version: string; destination: string } = input;
+    const { requestMetadata }: { requestMetadata: Object } = input;
+    const { feature }: { feature: string } = input;
+
+    const envThreshold = parseFloat(process.env.COMPARISON_TEST || '0');
+    const destThreshold = this.getTestThreshold(events[0].destination);
+    const compareTestThreshold = envThreshold * destThreshold;
+
+    if (
```
Comment (koladilip) [inline src/services/comparator.ts:31]: We need to invoke primary always and if compare test > 0 then only invoke secondary and compare

File: src/services/comparator.ts (hunk 1 above)
Comment (utsabc) [reply src/services/comparator.ts:19]: Made this generic pushed some changes can you check once?

File: src/services/comparator.ts (hunk 3 above)
Comment (utsabc) [reply src/services/comparator.ts]: This whole flow is async I have integrated in controller can you check once

File: src/services/comparator.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+import logger from '../logger';
+import { CommonUtils } from '../util/common';
+import { PlatformError } from '../v0/util/errorTypes';
+const NS_PER_SEC = 1e9;
+
+dotenv.config();
+
+export default class ComparatorService {
+  private static getTestThreshold(destination: Destination) {
+    return destination.DestinationDefinition?.Config['camparisonTestThreshold'] || 0;
```
Comment (koladilip) [inline src/services/comparator.ts:20]: this needs definition update

File: src/services/comparator.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    const { events, version, destination, requestMetadata, feature } = input;
+
+    try {
+      const primaryServiceName = process.env.COMPARISON_TEST_PRIMARY_SERVICE;
+      const secondaryServiceName = process.env.COMPARISON_TEST_SECONDARY_SERVICE;
+
+      const envThreshold = parseFloat(process.env.COMPARISON_TEST || '0');
+      const destThreshold = this.getTestThreshold(events[0].destination);
+      const compareTestThreshold = envThreshold * destThreshold;
+      if (
```
Comment (koladilip) [inline src/services/comparator.ts:33]: the primary is always called, and secondary is called conditionally based threshold

File: src/helpers/serviceSelector.ts [hunk 6]
```diff
@@ -70,16 +94,23 @@ export default class ServiceSelector {
...
-    switch (name) {
-      case INTEGRATION_SERVICE.CDK_V1_DEST:
-        return this.fetchCachedService(INTEGRATION_SERVICE.CDK_V1_DEST);
-      case INTEGRATION_SERVICE.CDK_V2_DEST:
-        return this.fetchCachedService(INTEGRATION_SERVICE.CDK_V2_DEST);
-      case INTEGRATION_SERVICE.NATIVE_DEST:
-        return this.fetchCachedService(INTEGRATION_SERVICE.NATIVE_DEST);
-      default:
-        throw new PlatformError('Invalid Service');
+  public static getDestinationService(
```
Comment (koladilip) [inline src/helpers/serviceSelector.ts:97]: document example destination definition config format for this

File: src/helpers/serviceSelector.ts [hunk 7]
```diff
@@ -53,11 +62,26 @@ export default class ServiceSelector {
...
+        throw new PlatformError('Invalid Service');
+    }
+  }
+
+  private static getPrimaryDestinationService(
     events: ProcessorTransformationRequest[] | RouterTransformationRequestData[],
   ): DestinationService {
     const destinationDefinitionConfig: Object = events[0].destination.DestinationDefinition.Config;
-    if (this.isCdkDestination(destinationDefinitionConfig)) {
+    if (this.isComparatorEnabled(destinationDefinitionConfig)) {
```
Comment (koladilip) [inline src/helpers/serviceSelector.ts:82]: is this needed?

File: src/helpers/serviceSelector.ts [hunk 8]
```diff
@@ -70,16 +94,23 @@ export default class ServiceSelector {
...
+    const destinationDefinitionConfig: Object = events[0].destination.DestinationDefinition.Config;
+    const primaryService = this.getPrimaryDestinationService(events);
+    if (this.isComparatorEnabled(destinationDefinitionConfig)) {
+      if (this.serviceMap.has(INTEGRATION_SERVICE.COMPARATOR)) {
+        return this.serviceMap.get(INTEGRATION_SERVICE.COMPARATOR);
+      }
+      const secondaryServiceName = this.getSecondaryServiceName(destinationDefinitionConfig);
+      const secondaryService = this.getDestinationServiceByName(secondaryServiceName);
+      const comparatorService = new ComparatorService(primaryService, secondaryService);
+      this.serviceMap.set(INTEGRATION_SERVICE.COMPARATOR, comparatorService);
```
Comment (koladilip) [inline src/helpers/serviceSelector.ts:109]: Maybe we shouldn't cache COMPARATOR as the combination can vary right?

File: src/helpers/serviceSelector.ts (hunk 7 above)
Comment (utsabc) [reply src/helpers/serviceSelector.ts:82]: yes removed

File: src/helpers/serviceSelector.ts (hunk 8 above)
Comment (utsabc) [reply src/helpers/serviceSelector.ts:109]: Ok destination level config right yes

File: src/helpers/serviceSelector.ts (hunk 6 above)
Comment (utsabc) [reply src/helpers/serviceSelector.ts:97]: You mean adding new type for config?

File: src/helpers/serviceSelector.ts (hunk 8 above)
Comment (utsabc) [reply src/helpers/serviceSelector.ts]: Added `destination_definition_id` as part of the `Key`

File: src/helpers/serviceSelector.ts (hunk 6 above)
Comment (koladilip) [reply src/helpers/serviceSelector.ts:95]: Yes, how the destination config should be created for this feature

**Review by koladilip — APPROVED** [review]

**Pull request merged**